src/factory.js

```diff
@@ -103,23 +103,22 @@ export default function plotComponentFactory(Plotly) {
     componentWillUpdate(nextProps) {
       this.unmounting = false;
 
-      if (nextProps.revision !== void 0 && nextProps.revision === this.props.revision) {
-        // if revision is set and unchanged, do nothing
-        return;
-      }
-
+      // frames *always* changes identity so fall back to check length only :(
       const numPrevFrames =
         this.props.frames && this.props.frames.length ? this.props.frames.length : 0;
       const numNextFrames =
         nextProps.frames && nextProps.frames.length ? nextProps.frames.length : 0;
-      if (
+
+      const figureChanged = !(
         nextProps.layout === this.props.layout &&
         nextProps.data === this.props.data &&
         nextProps.config === this.props.config &&
         numNextFrames === numPrevFrames
-      ) {
-        // prevent infinite loops when component is re-rendered after onUpdate
-        // frames *always* changes identity so fall back to check length only :(
+      );
+      const revisionDefined = nextProps.revision !== void 0;
+      const revisionChanged = nextProps.revision !== this.props.revision;
+
+      if (!figureChanged && (!revisionDefined || (revisionDefined && !revisionChanged))) {
```

Comment:
I think this condition should be `( (!revisionDefined && !figureChanged) || (revisionDefined && !revisionChanged) )`

Reply:
or even clearer:

```javascript

if(revisionDefined) {
    if (!revisionChanged) return;
}
else {
    if (!figureChanged) return;
}

```

Reply:
ooh you want clear?
```javascript
let revisionNotDefinedAndFigureNotChanged = !revisionDefined && !figureChanged;
let revisionDefinedAndRevisionNotChanged = revisionDefined && !figureChanged;
if (revisionNotDefinedAndFigureNotChanged || revisionDefinedAndRevisionNotChanged) { 
    return;
}
```

Reply:
Or may as well just keep going here
```javascript
let revisionNotDefinedAndFigureNotChangedOrRevisionDefinedAndRevisionNotChanged = revisionNotDefinedAndFigureNotChanged || revisionDefinedAndRevisionNotChanged;
if (revisionNotDefinedAndFigureNotChangedOrRevisionDefinedAndRevisionNotChanged) {
    return;
}
```

Reply:
https://github.com/plotly/react-plotly.js/pull/126#discussion_r258987078 is great :)

Reply:
DRY one-liner: `if( revisionDefined ? !revisionChanged : !figureChanged) return;`

Reply:
The one-liner is the clear winner here I think

Reply:
@nicolaskruchten Actually the one-liner doesn't take care of the case when 
revisionDefined = true
revisionChanged = false
figureChanged = true
That was the whole reason for this PR

Reply:
So now it's 
`(!figureChanged && (!revisionDefined || (revisionDefined && !revisionChanged)))`
vs
`((!revisionDefined && !figureChanged) || (revisionDefined && !revisionChanged))`

My original one works well for figure updates, because once the figure condition is tested, you know right away to drop the rest and start the update. I deprioritize revision stuff since it might be 🔪some time soon anyway. @nicolaskruchten's might be clearer though.

Reply:
You're right, I was still holding on to an earlier viewpoint. 

I'm 💃 on the current code actually. Sorry for allll the back and forth and thanks for continuing to push this thing in the right direction.

Reply:
It's just hard to read because it's defining the do-nothing case. It might be more readable to say `if(! (<define the positive case>) )...`

Reply:
Mine can be simplified to
`(!figureChanged && (!revisionDefined && !revisionChanged)))`
Forgot all my propositional calculus, tools to the rescue :)
http://logictools.org/
`-a | (a & -b) <=> (-a & -b)`

Reply:
Tools showed that @nicolaskruchten's and mine are not equivalent. Upon further inspection, `((!revisionDefined && !figureChanged) || (revisionDefined && !revisionChanged))` fails the same test as https://github.com/plotly/react-plotly.js/pull/126#discussion_r. So I'm gonna revert back to original. Any thoughts?

Reply:
Sorry, github didn't show the comments above until I refreshed, will merge...